policies/gac/networks.py

```diff
@@ -348,72 +354,75 @@ def train(self, transitions, value, gamma):
         return history
 
 
-    def __call__(self, states, actions):
-        x = tf.concat([states, actions], -1)
-        return self.model.predict(x)
 
 
-class Value():
+
+class Value(tf.Module):
 
     """
     Value network has the same architecture as Critic
     """
 
     def __init__(self, state_dim):
+        super(Value, self).__init__()
         self.model = _build_sequential_model(state_dim)
         self.state_dim = state_dim
 
+    def __call__(self, states):
+        return self.model.predict(states)
 
-    def train(self, transitions, action_sampler, actor, critic1, critic2, K):
+    def train(self, transitions, actor, critic1, critic2, action_samples = 64):
         """
         transitions is of type named tuple policies.policy_helpers.helpers.Transition
         action_sampler is of type policies.policy_helpers.helpers.ActionSampler
         """
 
-        """Each state needs K action samples"""
-        # [batch size , 1 , state dim]
-        states = tf.expand_dims(transitions.s, 1)
-        # [batch size , K , state dim]
-        states = tf.broadcast_to(states, [states.shape[0], K] + states.shape[2:])
-        # [batch size x K , state dim]
+        """Each state needs action_samples action samples"""
+
+        #TODO: this tiling part is pretty dangerous. NEED DOUBLE CHECK!
+
+        # originally, transitions.s is [batch_size , state_dim]
+        states = tf.expand_dims(transitions.s, 1) # [batch_size , 1 , state_dim]
+        states = tf.broadcast_to(states, [states.shape[0], action_samples] + states.shape[2:])
+        # [batch_size , action_samples , state dim]
         states = tf.reshape(states, [-1, self.state_dim])
+        # [batch_size x action_samples , state dim]
 
 
         """
         Line 13 of Algorithm 2.
         Sample actions from the actor network given current state and tau ~ U[0,1].
         """
-        actions = action_sampler.get_actions(actor, states)
+        batch_size = transitions.s.shape[0]
```

Comment:
The sampling is not only used here but also in many other places. So it's better to be just a line of code.

Reply:
If my understanding is correct, the authors leave this only because they want to add noise. If we don't want noise, it's nothing more than call the actor.